### PR TITLE
Wavefront Coverage Path Planning 

### PR DESCRIPTION
**Pull request by reso1**
Implemented the Distance/Path Transform Wavefront Coverage Path Planning Algorithm, as proposed in [Planning paths of complete coverage of an unstructured environment by a mobile robot](http://pinkwink.kr/attachment/cfile3.uf@1354654A4E8945BD13FE77.pdf) by Zelinsky et.al.

Tried to follow the coding style in GridBasedSweepCPP

### CODE REVIEW

**Comment by AtsushiSakai** [comment]
@reso1 Cool!! Thank you for your great PR!!

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+author: Todd Tang
+paper: Planning paths of complete coverage of an unstructured environment
+         by a mobile robot - Zelinsky et.al.
+link: http://pinkwink.kr/attachment/cfile3.uf@1354654A4E8945BD13FE77.pdf
+"""
+
+import os
+
+import cv2
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:12]: using opencv is not acceptable... Is it possible not to use it?

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+import numpy as np
+
+from scipy import ndimage
+
+
+def transform(
+    gridmap, src, distance_type='chessboard',
+    transform_type='path', alpha=0.01
+):
+    """transform
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:24]: Please fix code scanning warning.

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                if not is_visited[ni][nj] and ((ni-1)*ncols + nj) \
+                        not in calculated:
+                    queue.append((ni, nj))
+                    calculated.append((ni-1)*ncols + nj)
+
+    return T
+
+
+def wavefront(T, start, goal):
+    """wavefront
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:90]: Please fix code scanning warning

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        is_visited[i][j] = True
+
+        max_T = float('-inf')
+        imax = (-1, -1)
+        ilast = 0
+        for ilast in range(len(path)):
+            cur = path[-1-ilast]
+            for ci, cj in order:
+                ni, nj = cur[0] + ci, cur[1] + cj
+                if ni >= 0 and ni < nrows and nj >= 0 and nj < ncols \
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:132]: This long if condition should be a function which return boolean for readability

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+                 [1, 1], [1, -1], [-1, 1], [-1, -1]]
+    elif start[0] <= goal[0] and start[1] >= goal[1]:
+        order = [[-1, 0], [0, 1], [1, 0], [0, -1],
+                 [-1, 1], [-1, -1], [1, 1], [1, -1]]
+    elif start[0] >= goal[0] and start[1] <= goal[1]:
+        order = [[1, 0], [0, -1], [-1, 0], [0, 1],
+                 [1, -1], [-1, -1], [1, 1], [-1, 1]]
+    elif start[0] <= goal[0] and start[1] <= goal[1]:
+        order = [[-1, 0], [0, -1], [0, 1], [1, 0],
+                 [-1, -1], [-1, 1], [1, -1], [1, 1]]
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:114]: This block should be a function returning order for readability.

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
+"""
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:1]: Could you please add a unittest for this script like: https://github.com/AtsushiSakai/PythonRobotics/blob/master/tests/test_grid_based_sweep_coverage_path_planner.py

**Comment by reso1** [comment]
Thanks for the quick reply, the unittest will be committed soon:)

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        sys.exit('Unsupported transform type.')
+
+    # set obstacle T value to infinity
+    for i in range(nrows):
+        for j in range(ncols):
+            if gridmap[i][j] == 1.0:
+                T[i][j] = float('inf')
+    is_visited = np.zeros_like(T, dtype=bool)
+    is_visited[src[0], src[1]] = True
+    queue = [src]
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:68]: The name `queue` is unclear. What kind of queue is this?

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    cur = start
+    is_visited = np.zeros_like(T, dtype=bool)
+
+    while cur != goal:
+        i, j = cur
+        path.append((i, j))
+        is_visited[i][j] = True
+
+        max_T = float('-inf')
+        imax = (-1, -1)
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:137]: imax -> i_max

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+        ilast = 0
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:138]: ilast -> i_last

File: PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+    def is_valid_neighbor(i, j):
+        is_i_valid_bounded = i >= 0 and i < nrows
+        is_j_valid_bounded = j >= 0 and j < ncols
+        if is_i_valid_bounded and is_j_valid_bounded:
+            return not is_visited[i][j] and T[i][j] != float('inf')
+        return False
+
+    inc_order = get_search_order_increment(start, goal)
+
+    cur = start
```
Comment (AtsushiSakai) [inline PathPlanning/WavefrontCPP/wavefront_coverage_path_planner.py:128]: `cur` is current? if so, please use `current_position` or something.

**Comment by reso1** [comment]
Modified the codes following the code review, thanks for the code review again :)

**Comment by AtsushiSakai** [comment]
@reso1 Thank you!!. I will merge it.

**Pull request merged**